### DISCUSSION TITLE
Streaming response while streaming request (full-duplex)

### DISCUSSION THREAD

**Discussion by hrist0stoichev** [fastapi/fastapi#13126] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
from fastapi.responses import StreamingResponse

app = FastAPI()

@app.post("/stream-uppercase")
async def stream_uppercase(request: Request):
    """
    Endpoint that reads the request body as a stream,
    extracts only uppercase letters, and streams them back.
    """

    async def uppercase_generator():
        async for chunk in request.stream():
            text = chunk.decode("utf-8", errors="ignore")
            uppercase_only = "".join([ch for ch in text if ch.isupper()])
            if uppercase_only:
                yield uppercase_only

    return StreamingResponse(uppercase_generator(), media_type="text/plain")
```

### Description

```
echo "Hello WORLD" > test.txt
curl -X POST --data-binary @test.txt http://localhost:8000/stream-uppercase
```

When I call this I expect that the uppercase letters would be streamed back to me. However, what ends up happening is that the program hangs on line `async for chunk in request.stream()`. I added a print statement inside the loop and I can't anything being printed on the terminal.

### Operating System

macOS

### FastAPI Version

0.115.6

### Pydantic Version

1.10.19

### Python Version

3.10.11

**Comment by Kludex** [comment, marked as answer]
That's correct.

The latest Starlette with Uvicorn supports this.

Unfortunately FastAPI didn't bump the minimum version of Starlette, so you'd need to wait a bit.

**Reply by hrist0stoichev** [reply]
Thank you! I was experimenting with `hypercorn` thinking I need HTTP2 to make this work (although I read HTTP 1.1 supports full-duplex). Since `Uvicorn` supports only HTTP1.1, does you answer imply that HTTP2 is not needed?

**Reply by Kludex** [reply]
HTTP2 is not needed.

**Reply by hrist0stoichev** [reply]
fastapi just had a new release allowing for the latest version of Starlette but I still can't make this run...

**Reply by hrist0stoichev** [reply]
OK, so I thought I might need to start reading the stream **before** sending the request and decided to reimplement it using a queue:

```
@app.post("/stream-uppercase-2")
async def stream_uppercase_2(request: Request):
    """
    Endpoint that reads the request body as a stream,
    extracts only uppercase letters, and streams them back using a queue.
    """
    queue = asyncio.Queue()
    
    async def producer():
        try:
            async for chunk in request.stream():
                print("reading chunk with length", len(chunk))
                text = chunk.decode("utf-8", errors="ignore")
                uppercase_only = "".join([ch for ch in text if ch.isupper()])
                if uppercase_only:
                    await queue.put(uppercase_only)
            await queue.put(None)  # Signal end of stream
        except Exception as e:
            await queue.put(None)
            raise e

    async def consumer():
        while True:
            chunk = await queue.get()
            if chunk is None:  # Check for end of stream
                break
            print("writing chunk with length", len(chunk))
            yield chunk
            queue.task_done()

    # Start producer task in background
    asyncio.create_task(producer())

    # Sleep so that we start reading the input stream before sending the response
    await asyncio.sleep(0.01)

    return StreamingResponse(consumer())
```

And this now works! Importantly, I had to add the sleep, otherwise it's not working. This made me think that my assumption was correct and, indeed, the solution was to start reading the input stream before sending the request. So, I tried "fixing" my original code like that:

```
@app.post("/stream-uppercase")
async def stream_uppercase(request: Request):
    """
    Endpoint that reads the request body as a stream,
    extracts only uppercase letters, and streams them back.
    """
    gen = request.stream()

    # Read the first chunk only so that we start reading the input stream before sending the response.
    async for chunk in gen:
        text = chunk.decode("utf-8", errors="ignore")
        print("reading chunk with length", len(text))
        break

    async def uppercase_generator():
        async for chunk in gen:
            print("reading chunk from generator with length", len(chunk))
            text = chunk.decode("utf-8", errors="ignore")
            uppercase_only = "".join([ch for ch in text if ch.isupper()])
            if uppercase_only:
                print("writing chunk with length", len(uppercase_only))
                yield uppercase_only

    return StreamingResponse(uppercase_generator())
```

However, this still hangs and I never see the "reading chunk from generator with length" message.

Importantly, to simulate streaming on localhost I use
```
curl -X POST --data-binary @test.txt http://localhost:8000/stream-uppercase --limit-rate 1k --no-buffer
```

**Reply by Kludex** [reply]
If you are in the latest uvicorn and starlette, it should work without this queue.

**Reply by hrist0stoichev** [reply]
I am on latest uvicorn (0.34.0), fastapi (0.115.7), and starlette (0.45.2). I'm running the server using `uvicorn main:app --host 0.0.0.0 --port 8000` and it's hanging. Not sure what I might be doing wrong.

**Reply by Kludex** [reply]
Hmmm... I thought it was implemented. My bad, I think I implemented and discarded the branch.

Uvicorn doesn't support it yet. It may actually work with a previous version of uvicorn e.g. 0.30.0 but it's not reliable on disconnection.

**Reply by hrist0stoichev** [reply]
What do you think about the Queue implementation in that case? Do you think it's reliable?

**Reply by caffeinism** [reply]
@Kludex Is it against the protocol of HTTP to return a response to get the entire body of a request? (I'm just curious because I haven't found anything about this.) Also, does uvicorn have any plans to support this behavior, or would it be accepted if I presented a PR like this?
